### PR TITLE
.Net: Add File Id Reference Handling to Experimental Agent Package so that User Messages can be related to file Ids in OpenAI Storage

### PR DESCRIPTION
**Pull request by Ryan-Britton**
### Motivation and Context

added the ability to reference file ids when creating a user message in the experimental agents package

image

### Description

added a nullable array of file ids as a parameter on the AddUserMessageAsync and InvokeAsync methods so that it did not break existing tests/examples. Added the file_ids property to the POST payload to the OpenAI Assistants API.

### Contribution Checklist

- [Y ] The code builds clean without any errors or warnings
- [ Y] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [Y ] All unit tests pass, and I have added new tests where possible
- [ Y] I didn't break anyone :smile:

### CODE REVIEW

**Comment by Ryan-Britton** [comment]
@microsoft-github-policy-service agree [company="Wet Ink Corporation"]

**Comment by Ryan-Britton** [comment]
@microsoft-github-policy-service agree

**Comment by crickman** [comment]
@Ryan-Britton - Can you look at this change I published last week: https://github.com/microsoft/semantic-kernel/pull/5113

**Comment by Ryan-Britton** [comment]
they're definitely related - looking at your changes, I see that you're using the file_id paramter for file references on chat messages in #5113 - but in your use case it seems to be looking specifically at images (where I was using the file_ids parameter to refer to any file type in OpenAI storage as a context for the chat). My thought during the quick review was that the naming of some of your changes to align it "image types" specifically, may obfuscate the fact that file_ids can point at non-image types as well?

**Comment by crickman** [comment]
Made a couple small tweaks, including adding a mode to `Example75_AgentTools` that pass the fileId via the thread.

Can you take a look?

**Review by crickman — APPROVED** [review]

**Review by crickman — APPROVED** [review]

File: dotnet/src/Experimental/Agents/Internal/Agent.cs [hunk 1]
```diff
@@ -220,20 +221,22 @@ public async Task DeleteAsync(CancellationToken cancellationToken = default)
...
     /// <param name="input">The user input</param>
     /// <param name="arguments">Arguments for parameterized instructions</param>
+    /// <param name="fileIds">an array of up to 10 file ids to reference for the message</param>
     /// <param name="cancellationToken">A cancellation token.</param>
     /// <returns>An agent response (<see cref="AgentResponse"/></returns>
     private async Task<AgentResponse> AskAsync(
         [Description("The user message provided to the agent.")]
         string input,
         KernelArguments arguments,
+        string[]? fileIds = null,
```
Comment (SergeyMenshykh) [inline dotnet/src/Experimental/Agents/Internal/Agent.cs:231]: "IEnumerable<string>? fileIds" to be aligned the other changes?

File: dotnet/src/Experimental/Agents/Internal/ChatThread.cs [hunk 2]
```diff
@@ -50,33 +50,31 @@ public static async Task<IAgentThread> GetAsync(OpenAIRestContext restContext, s
...
+    public async Task<IChatMessage> AddUserMessageAsync(string message, IEnumerable<string>? fileIds = null, CancellationToken cancellationToken = default)
     {
         this.ThrowIfDeleted();
 
-        var messageModel =
-            await this._restContext.CreateUserTextMessageAsync(
-                this.Id,
-                message,
-                cancellationToken).ConfigureAwait(false);
+        var messagemodel = fileIds == null
```
Comment (SergeyMenshykh) [inline dotnet/src/Experimental/Agents/Internal/ChatThread.cs:57]: Can't it just be simplified to - "var messageModel = await this._restContext.CreateUserTextMessageAsync(this.Id, message, **fileIds,**, cancellationToken).ConfigureAwait(false);" ?

**Review by SergeyMenshykh — APPROVED** [review]

File: dotnet/src/Experimental/Agents/Internal/ChatThread.cs (hunk 2 above)
Comment (crickman) [reply dotnet/src/Experimental/Agents/Internal/ChatThread.cs:57]: Yes, both comments are a side-effect of my edits on top of Ryan's PR.

File: dotnet/src/Experimental/Agents/Internal/Agent.cs (hunk 1 above)
Comment (crickman) [reply dotnet/src/Experimental/Agents/Internal/Agent.cs:231]: This is a a kernel-function and private.  I'm not sure the principal of alignment is impactful here (as much as I'm in favor of that).  I'd like to keep this simple / as-is for now to ensure functionality.

**Review by crickman — APPROVED** [review]

File: dotnet/src/Experimental/Agents/Internal/ChatThread.cs [hunk 3]
```diff
@@ -50,33 +50,29 @@ public static async Task<IAgentThread> GetAsync(OpenAIRestContext restContext, s
...
+    public async Task<IChatMessage> AddUserMessageAsync(string message, IEnumerable<string>? fileIds = null, CancellationToken cancellationToken = default)
     {
         this.ThrowIfDeleted();
 
-        var messageModel =
-            await this._restContext.CreateUserTextMessageAsync(
-                this.Id,
-                message,
-                cancellationToken).ConfigureAwait(false);
+        var messagemodel = await this._restContext.CreateUserTextMessageAsync(this.Id, message, fileIds: null, cancellationToken).ConfigureAwait(false);
```
Comment (glahaye) [inline dotnet/src/Experimental/Agents/Internal/ChatThread.cs:57]: You set fileIds to null all the time here instead of passing the fileIds parameter you got above?

Comment (crickman) [reply dotnet/src/Experimental/Agents/Internal/ChatThread.cs:57]: Right, my bad

File: dotnet/src/Experimental/Agents/Internal/ChatThread.cs [hunk 4]
```diff
@@ -50,33 +50,29 @@ public static async Task<IAgentThread> GetAsync(OpenAIRestContext restContext, s
...
 
-        return new ChatMessage(messageModel);
+        return new ChatMessage(messagemodel);
```
Comment (glahaye) [inline dotnet/src/Experimental/Agents/Internal/ChatThread.cs:59]: Nit: why did you change the nameCasing to be incorrect (for C#) now?

File: dotnet/samples/KernelSyntaxExamples/Example75_AgentTools.cs [hunk 5]
```diff
@@ -91,6 +96,7 @@ public async Task RunRetrievalToolAsync()
                 new OpenAIFileUploadExecutionSettings("travelinfo.txt", OpenAIFilePurpose.Assistants));
 
         var fileId = result.Id;
+        this.WriteLine($"! {fileId}");
```
Comment (glahaye) [inline dotnet/samples/KernelSyntaxExamples/Example75_AgentTools.cs:99]: Did you mean to keep this in?

File: dotnet/src/Experimental/Agents/Internal/ChatThread.cs (hunk 4 above)
Comment (crickman) [reply dotnet/src/Experimental/Agents/Internal/ChatThread.cs:59]: As you may be familiar, sometimes visual studio does things. ;)

File: dotnet/samples/KernelSyntaxExamples/Example75_AgentTools.cs [hunk 6]
```diff
@@ -140,7 +159,7 @@ public async Task RunRetrievalToolAsync()
 
         async Task InvokeAgentAsync(IAgent agent, string question)
         {
-            await foreach (var message in agent.InvokeAsync(question))
+            await foreach (var message in agent.InvokeAsync(question, null, fileIds))
```
Comment (glahaye) [inline dotnet/samples/KernelSyntaxExamples/Example75_AgentTools.cs:162]: You pass null for KernelArguments but InvokeAsync() takes in "KernelArguments", not "KernelArguments?".   Maybe InvokeAsync() should be made to take in a KernelArguments? ???

File: dotnet/samples/KernelSyntaxExamples/Example75_AgentTools.cs (hunk 5 above)
Comment (crickman) [reply dotnet/samples/KernelSyntaxExamples/Example75_AgentTools.cs:99]: Yes

File: dotnet/samples/KernelSyntaxExamples/Example75_AgentTools.cs (hunk 6 above)
Comment (glahaye) [reply dotnet/samples/KernelSyntaxExamples/Example75_AgentTools.cs:162]: Or maybe I'm looking at the wrong InvokeAsync()?

Comment (crickman) [reply dotnet/samples/KernelSyntaxExamples/Example75_AgentTools.cs:162]: image

Comment (crickman) [reply dotnet/samples/KernelSyntaxExamples/Example75_AgentTools.cs:162]: I think it does accept null.  I'm confused.

Comment (glahaye) [reply dotnet/samples/KernelSyntaxExamples/Example75_AgentTools.cs:162]: I must be hallucinating again.  Never mind this :)

**Review by glahaye — APPROVED** [review]
